kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,164 @@
+            }
+
+            Map<String, String> data = cm.getData();
+            kieServerState = (KieServerState) xs.fromXML(data.get(CFG_MAP_DATA_KEY));
+
+        } catch (IllegalStateException ise) {
+            logger.error("Loading KieServerState failed.", ise);
+            throw ise;
+        } catch (Exception e) {
+            logger.error("Loading KieServerState failed.", e);
+        }
+
+        if (kieServerState == null || !retrieveKieServerId(kieServerState).equals(serverId)) {
+            throw new IllegalStateException("Invalid KieServerId: " + serverId + ", or inconsistent state data.");
+        }
+
+        return kieServerState;
+    }
+
+    @Override
+    public boolean isKieServerReady() {
+        return KieServerLocator.getInstance().isKieServerReady();
+    }
+
+    protected String retrieveKieServerId(KieServerState kieServerState) {
+        String kssServerId = null;
+
+        try {
+            kssServerId = kieServerState.getConfiguration().getConfigItemValue(KieServerConstants.KIE_SERVER_ID);
+        } catch (Exception e) {
+            logger.error("Failed to retrieve server id from KieServerState", e);
+        }
+
+        if (kssServerId == null || kssServerId.length() == 0) {
+            throw new IllegalArgumentException("Invalid KieServerId: Can not be null or empty.");
+        }
+        return kssServerId;
+    }
+
+    protected void createOrUpdate(String serverId, KieServerState kieServerState, boolean isUpdate) {
+        try (KubernetesClient client = createKubernetesClient()) {
+
+            ConfigMap cm = client.configMaps().withName(serverId).get();
+
+            if (isUpdate && cm == null) {
+                throw new IllegalStateException("KieServerState ConfigMap must exist before update.");
+            } else {
+                cm = client.configMaps().createOrReplace(new ConfigMapBuilder()
+                                                                               .withNewMetadata()
+                                                                               .withName(serverId)
+                                                                               .endMetadata().build());
+            }
+
+            String stateXML = xs.toXML(kieServerState);
+            ObjectMeta md = cm.getMetadata();
+
+            Map<String, String> ann = md.getAnnotations() == null ? new ConcurrentHashMap<>() : md.getAnnotations();
+            md.setAnnotations(ann);
+            ann.put(STATE_CHANGE_TIMESTAMP,
+                    ZonedDateTime.now().format(DateTimeFormatter.ISO_INSTANT));
+
+            if (isKieServerReady()) {
+                ann.put(ROLLOUT_REQUIRED, "true");
+            }
+
+            cm.setData(Collections.singletonMap(CFG_MAP_DATA_KEY, stateXML));
+            client.configMaps().createOrReplace(cm);
```

Comment:
Does it mean that the config map is updated even without triggering rollout?

Reply:
It is for backward compatibility. During the KieServer initialization phase, there are calls to Repo.store to persist settings from environment variables and system properties.

Reply:
So Kie server should store there properties/variables which aren't defined by user/controller during creation of config map?